gcovr/gcov.py

```diff
@@ -172,6 +174,9 @@ def guess_source_file_name(
             gcovname, currdir, root_dir, starting_dir, obj_dir, source_fname
         )
 
+    if use_canonical_paths:
+        fname = os.path.realpath(fname)
+
```

Comment:
Often, `fname` will already be a canonical path. I think it would be correct to *always* use `fname = os.path.realpath(fname)` here. Removing such conditionals simplifies testing.

Reply:
Shall I change the code appropriately then?

Reply:
I've checked the surrounding code. I think applying an unconditional `realpath()` in that location is neither harmful nor necessary. So I'd leave it in, and see if any tests fail. But there are many uses of `normpath()` or `abspath()` that should probably be audited.

Reply:
Without this line, gcovr failed to match the files correctly in my environment. So removing it makes no sense. I'll turn it unconditional instead.

Reply:
@latk In my tests with substitutions the test `filter-relative-lib-from-unfiltered-tracefile` fails with this. I think the test data is incorrect here because the filters for the tracefile are:
```
Filters for --root: (1)
- re.compile('^\\/gcovr\\/gcovr\\/tests\\/filter\\-relative\\-lib\\-from\\-unfiltered\\-tracefile\\/project\\/')
Filters for --filter: (1)
- DirectoryPrefixFilter(\/gcovr\/gcovr\/tests\/filter\-relative\-lib\-from\-unfiltered\-tracefile\/project\/)
```
and the file `yes.cpp` is filtered out in the tracefile:
```
  fname        /gcovr/gcovr/tests/filter-relative-lib-from-unfiltered-tracefile/external-library/src/yes.cpp
Parsing coverage data for file /gcovr/gcovr/tests/filter-relative-lib-from-unfiltered-tracefile/external-library/src/yes.cpp
  Filtering coverage data for file /gcovr/gcovr/tests/filter-relative-lib-from-unfiltered-tracefile/external-library/src/yes.cpp
```

Reply:
Indeed, the test seems incorrect. In the gcovr invocation that creates the JSON file, I think there should be a `--root ..` so that the libraries aren't prematurely excluded. No idea what we were thinking back then. I'm glad the work on this PR has discovered that bug.

Reply:
I've checked the issue again and it seems that it came from to much realpath calls. So nothing to do at the moment.